python/cudf/cudf/core/column/numerical.py

```diff
@@ -274,38 +273,18 @@ def nans_to_nulls(self: NumericalColumn) -> NumericalColumn:
 
     def normalize_binop_value(
         self, other: ScalarLike
-    ) -> Union[ColumnBase, ScalarLike]:
+    ) -> Union[ColumnBase, cudf.Scalar]:
         if isinstance(other, ColumnBase):
             if not isinstance(other, NumericalColumn):
                 return NotImplemented
             return other
         if isinstance(other, cudf.Scalar):
-            if self.dtype == other.dtype:
-                return other
-            # expensive device-host transfer just to
-            # adjust the dtype
-            other = other.value
-        other_dtype = np.min_scalar_type(other)
-        if other_dtype.kind in {"b", "i", "u", "f"}:
-            if isinstance(other, cudf.Scalar):
-                return other
-            other_dtype = np.promote_types(self.dtype, other_dtype)
-            if other_dtype == np.dtype("float16"):
-                other_dtype = cudf.dtype("float32")
-                other = other_dtype.type(other)
-            if self.dtype.kind == "b":
-                other_dtype = min_signed_type(other)
-            if np.isscalar(other):
-                return cudf.dtype(other_dtype).type(other)
-            else:
-                ary = full(len(self), other, dtype=other_dtype)
-                return column.build_column(
-                    data=as_buffer(ary),
-                    dtype=ary.dtype,
-                    mask=self.mask,
-                )
-        else:
-            return NotImplemented
+            # Rely on C++ type promotion rules
```

Comment:
Are c++ type promotion rules the same as that of pandas?

Reply:
no

Reply:
see also #12039 (in particular @seberg's comment)

Reply:
Given the legitimate test failures, I need to think a bit harder about what to do here.

Reply:
That was a journey...; I think I got it right this time. Along the way I also found #12092 (fixed here) and #5938 (also fixed here).